### DISCUSSION TITLE
NER certainties

### DISCUSSION THREAD

**Discussion by GA-Ramlov** [explosion/spaCy#8942] [Help: Coding & Implementations] [answered]
Hi,

I'm developing a NER model using spacy and prodigy that I want to use for prediction on a larger unlabeled dataset.

While predicting the entities I would like to know the certainty of these predictions. However I cant figure out how to get the certainty of the model on each prediction/entity extraction it makes.

Hope someone can guide me in the correct direction.

Best
Jacob

**Comment by polm** [comment, marked as answer]
Please see #5917. This feature was recently added in #8855 and hasn't been released yet but will be soon (in the next version after 3.1.1). However, it also only works with the spancat component, not the default NER component. 

Note these are usually called "confidence" scores, not "certainties".

**Reply by wjbmattingly** [reply]
This is a great new feature!

**Reply by GA-Ramlov** [reply]
So if we have a text like:

"Just reviewed the new PS5 and that is super awesome"

And I have trained a model to get product names it would extract: "PS5" and then I would like to get the confidence score on that extraction. And also if the PS5 was like Playstation 5 then it will extract "Playstation 5" and get the confidence.

I have the model I just would like the confidence

Is it correctly understood that that would be the functionality?

Best
Jacob

**Reply by mbrunecky** [reply]
Yes, for each SPAN prediction you will be able to get the 'confidence'. Unfortunately, it is limited to the 'spancat' model - does not apply to 'ner' or any transformer models. But you can train your model using the 'spancat' pipeline instead of 'ner' - if you use Spacy command line 'train' then you use the same training data and a different config.cfg, The prediction query code (using SpanCategorzer) is slightly different: after (implied) predict() you look at doc.spans instead of doc.ents, and confidences are returned as an attribute carrying a list of confidences in the same order as the doc.spans. I am using the early version of the code (patch), and returned confidences make sense.
That said, the only example how to configure config.cfg for SpanCategorizer I found in the spancat demo.

```
            doc = nlp(text)
            # The key is specified in config.cfg
            for key in doc.spans:
                # Getting the scores (#8855 incorporated into spancat.py) attached as attr "scores"
                spans  = doc.spans[key]
                scores = spans.attrs.get("scores")
                print("scores", scores, "spans", spans)
                for i in range(len(spans)):
                    span  = spans[i]
                    score = 1.0 if scores is None else scores[i]
 
```

**Reply by polm** [reply]
Just a note - the spancat works with Transformers just fine. It is not included in the pretrained pipelines, including the Transformers ones, but in your own pipelines it can be used after a tok2vec or Transformer without issue.

**Reply by GA-Ramlov** [reply]
Okay. I'm training my model on XLM-Roberta as I have multiple languages. I guess that would work as well?

I have tried to search for the different drawbacks and benefits of the spancat versus ner but was not able to find any really descriptive ones.

Do you have a few comments on the pros and cons / difference between spancat and ner?

Best
Jacob

**Reply by polm** [reply]
You can use any model that works with spacy-transformers, or any of spaCy's tok2vec models. They're all just different sources of input to the spancat. 

For notes on the spancat vs NER see #8930. Basically NER should be more accurate because it makes simplifying assumptions, so you shouldn't use the spancat unless you specifically need a feature it has.

**Reply by mbrunecky** [reply]
Are you saying that I can use 'spancat' (to get individual span/label confidences) along with the 'transformer' or 'ner' models?
I mean using pipeline:
```
pipeline='transformer','spancat'
pipeline='vord2vec','ner','spancat'
```
???
Could I train my model using 'transformer' (or 'ner') only and then, for my predictions, add 'spancat' to the pipeline?
(how does the spancat 'input' mapping work?)

Besides getting confidences, I see the spancat value in adding my own 'span' constraints. I modified the 'span suggester' to skip any span NOT starting in uppercase or digit (my spans are names), and it helped both the training scores but mainly the training/validation performance (having to look at barely 30% of the spans) and memory utilization (I was running out on 64G machine). That said, reading the original span suggester code is ... not easy.

**Reply by polm** [reply]
> Are you saying that I can use 'spancat' (to get individual span/label confidences) along with the 'transformer' or 'ner' models?

Using it with Transformers and using it with NER are two very different things. 

Transformers or tok2vec components take input tokens and turn them into embeddings. This is a necessary first step for basically all other components, and you can use these components interchangeably. So you can use a Transformer as input to spancat.

NER is a different layer that gives output similar to spancat. They can't provide input to each other by default, though you could make a custom component that uses NER output to get proposed spans for spancat, for example. 

Note that a Transformer **component** is not the same thing as the Transformer **pipeline**, like `en_core_web_trf`. That pipeline is a pipeline that uses a Transformer component instead of tok2vec and is pretrained.

So for your specific examples:

    pipeline='transformer','spancat'

This is totally normal.

    pipeline='vord2vec','ner','spancat'

With the default setup, you can do this (assuming `vord2vec` is `tok2vec` - note `tok2vec` and `word2vec` are different), but by default the NER and spancat won't interact directly. So for example the spancat won't automatically give you NER scores or something.

> Besides getting confidences, I see the spancat value in adding my own 'span' constraints. 

This is a great reason to use the spancat instead of the NER component. For many NER tasks though, like recognizing a category of proper nouns like person, location, or brand names, it's not relevant.
